File: java/src/main/java/cn/leetcode/rejectsampling/$478_GenerateRandomPointInACircle.java
package cn.leetcode.rejectsampling;

public class $478_GenerateRandomPointInACircle {
    double rad, xc, yc;

    public $478_GenerateRandomPointInACircle(double radius, double x_center, double y_center) {
        rad = radius;
        xc = x_center;
        yc = y_center;
    }

    public double[] randPoint() {
        double x0 = xc - rad;
        double y0 = yc - rad;

        while (true) {
            double xg = x0 + Math.random() * rad * 2;
            double yg = y0 + Math.random() * rad * 2;
            if (Math.sqrt(Math.pow((xg - xc), 2) + Math.pow((yg - yc), 2)) <= rad)
                return new double[]{xg, yg};
        }
    }

}

File: python/rejection_sampling/$478_GenerateRandomPointInACircle.py
from math import sqrt

import math
import random
from typing import List


class Solution:

    def __init__(self, radius: float, x_center: float, y_center: float):
        self.radius = radius
        self.x_center = x_center
        self.y_center = y_center

    def randPoint(self) -> List[float]:
        theta = random.uniform(0, 2 * math.pi)  # 生成一个[0, 2pi]的角度theta
        r = sqrt(random.uniform(0, 1)) * self.radius  # 随机生成一个[0, 1]范围的值乘以radius得到[0,radius]的r
        return [self.x_center + r * math.sin(theta), self.y_center + r * math.cos(theta)]  # 代入公式

# Your Solution object will be instantiated and called as such:
# obj = Solution(radius, x_center, y_center)
# param_1 = obj.randPoint()
